pkg/cloudprovider/openstack.go

```diff
@@ -0,0 +1,848 @@
+		return nil, fmt.Errorf("networkID '%s' is not a valid UUID", networkID)
+	}
+
+	opts := neutronsubnets.ListOpts{NetworkID: networkID}
+	pager := neutronsubnets.List(o.neutronClient, opts)
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		subnetList, err := neutronsubnets.ExtractSubnets(page)
+		if err != nil {
+			return false, err
+		}
+		subnets = append(subnets, subnetList...)
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return subnets, nil
+}
+
+// getNovaServer gets the nova server with ID == <serverID>.
+func (o *OpenStack) getNovaServer(serverID string) (*novaservers.Server, error) {
+	if _, err := uuid.Parse(serverID); err != nil {
+		return nil, fmt.Errorf("serverID '%s' is not a valid UUID", serverID)
+	}
+
+	server, err := novaservers.Get(o.novaClient, serverID).Extract()
+	if err != nil {
+		return nil, err
+	}
+	return server, nil
+}
+
+// listNovaServerPorts lists all ports that are attached to the provided nova server
+// with ID == <serverID>.
+func (o *OpenStack) listNovaServerPorts(serverID string) ([]neutronports.Port, error) {
+	var err error
+	var serverPorts []neutronports.Port
+
+	if _, err := uuid.Parse(serverID); err != nil {
+		return nil, fmt.Errorf("serverID '%s' is not a valid UUID", serverID)
+	}
+
+	portListOpts := neutronports.ListOpts{
+		DeviceID: serverID,
+	}
+
+	pager := neutronports.List(o.neutronClient, portListOpts)
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		portList, err := neutronports.ExtractPorts(page)
+		if err != nil {
+			return false, err
+		}
+		serverPorts = append(serverPorts, portList...)
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return serverPorts, nil
+}
+
+// isIPAddressAllowedOnNeutronPort returns true if the given IP address can be found inside the
+// list of allowed_address_pairs for this port.
+func isIPAddressAllowedOnNeutronPort(p neutronports.Port, ip net.IP) bool {
+	for _, aap := range p.AllowedAddressPairs {
+		if ip.Equal(net.ParseIP(aap.IPAddress)) {
+			return true
+		}
+
+	}
+	return false
+}
+
+// hasNodeName will report true if the given description can be parsed correctly and
+// if given nodeName can be found inside the description.
+// If the nodeName cannot be found, return false, if the JSON cannot be parsed, return
+// an error.
+func hasNodeName(description string, nodeName string) (bool, error) {
+	d := EgressIPDescription{}
+	err := json.Unmarshal([]byte(description), &d)
+	if err != nil {
+		return false, err
+	}
+	return d.Type == egressIPTag && d.NodeName == nodeName, nil
+}
+
+// getNovaServerIDFromProviderID extracts the nova server ID from the given providerID.
+func getNovaServerIDFromProviderID(providerID string) (string, error) {
+	serverID := strings.TrimPrefix(providerID, openstackProviderPrefix)
+	if _, err := uuid.Parse(serverID); err != nil {
+		return "", fmt.Errorf("Cannot parse valid nova server ID from providerId '%s'", providerID)
```

Comment:
NIT: error type should not start with Capital letters, right?

Reply:
done